flink-runtime/src/main/java/org/apache/flink/runtime/io/network/NetworkEnvironment.java

```diff
@@ -202,28 +233,97 @@ public TaskKvStateRegistry createKvStateTaskRegistry(JobID jobId, JobVertexID jo
 		return kvStateRegistry.createTaskRegistry(jobId, jobVertexId);
 	}
 
+	public ResultPartition[] getResultPartitionWriters(String taskId) {
+		return allPartitions.get(taskId);
+	}
+
+	public SingleInputGate[] getInputGates(String taskId) {
+		return allInputGates.get(taskId);
+	}
+
 	// --------------------------------------------------------------------------------------------
-	//  Task operations
+	//  Create Output Writers and Input Readers
 	// --------------------------------------------------------------------------------------------
 
-	public void registerTask(Task task) throws IOException {
-		final ResultPartition[] producedPartitions = task.getProducedPartitions();
+	public ResultPartitionWriter[] createResultPartitionWriters(
+			String taskId,
+			JobID jobId,
+			ExecutionAttemptID executionId,
+			TaskActions taskActions,
+			ResultPartitionConsumableNotifier partitionConsumableNotifier,
+			TaskIOMetricGroup metrics,
+			Collection<ResultPartitionDeploymentDescriptor> resultPartitionDeploymentDescriptors) throws IOException {
+
+		if (isShutdown) {
+			throw new IllegalStateException("NetworkEnvironment is shut down");
+		}
 
-		synchronized (lock) {
-			if (isShutdown) {
-				throw new IllegalStateException("NetworkEnvironment is shut down");
-			}
+		ResultPartition[] parititons = new ResultPartition[resultPartitionDeploymentDescriptors.size()];
+		int counter = 0;
+		for (ResultPartitionDeploymentDescriptor rpdd : resultPartitionDeploymentDescriptors) {
+			parititons[counter] = new ResultPartition(
+				taskId,
+				taskActions,
+				jobId,
+				new ResultPartitionID(rpdd.getPartitionId(), executionId),
+				rpdd.getPartitionType(),
+				rpdd.getNumberOfSubpartitions(),
+				rpdd.getMaxParallelism(),
+				resultPartitionManager,
+				partitionConsumableNotifier,
+				ioManager,
+				rpdd.sendScheduleOrUpdateConsumersMessage());
+
+			setupPartition(parititons[counter]);
+			counter++;
+		}
 
-			for (final ResultPartition partition : producedPartitions) {
-				setupPartition(partition);
+		allPartitions.put(taskId, parititons);
+
+		metrics.initializeOutputBufferMetrics(parititons);
```

Comment:
Initially, we thought that ShuffleService would create separately partitions/gates.
Now it seems that metrics and partition manager keep partitions/gates by task.
It might change then ShuffleService interface.

As one of steps, we should probably refactor out netty specifics from `TaskIOMetricGroup` and register those specific metrics inside netty ShuffleService.

Reply:
Yes, this change is not considered before. The current `IOMetrics` relies on specific `ResultPartition` not regular `ResultPartitionWriter` interface. The final form of ShuffleService interface might include `{taskIdentifier(executionId or taskId), RPDD, metrics}`.

It would seem better if we could make `IOMetricGroup` created inside netty `ShuffleService` to make interface clean. But I have not thought through whether it is feasible to do like that.
From another aspect, if we pass `IOMetricGroup` to the interface as now, it might also seem reasonable because the regular metric framework could be also suitable for other writers. I referred to `ShuffleManager#getReader|getWriter` in spark before, and it also has the similar metric component in the interface.

After the metrics initialization is migrated into `ShuffleService` stack as now, `Task` can only reference with regular `ResultPartitionWriter` and `InputGate` arrays to solve some issues in #7549 .

Do you have further suggestions for metrics?

Reply:
Actually. looking more into `ResultPartitionManager`, we use indexing by execution id eventually only from task and task already knows its partitions/gates. Maybe, if we move all these partition/gate release/close for-loops from `NetworkEnvironment.unregisterTask` and `ResultPartitionManager.releasePartitionsProducedBy` to `Task`, this would help further to simplify future ShuffleService. If needed, netty `ResultPartition` could remove itself from `ResultPartitionManager` in `release()` method. This would be one step forward to get rid of "by execution id" indexing in `NetworkEnvironment`/`ResultPartitionManager`.

Then we have only metrics with "task to partitions/gates" dependency. Here, it does not seem to be trivial. In general, looking into `TaskIOMetricGroup.initializeOutputBufferMetrics`, `createResultPartitionWriters/gates` could just accept `ProxyMetricGroup<?>` instead of `TaskIOMetricGroup`. Then `initializeOutputBufferMetrics` could go into `NetworkEnvironment` using `ProxyMetricGroup.addGroup` with all netty specific gauge classes.

Thinking more about register "per task" VS "per partition/gate", "per task" is actually more restrictive and we do not need Shuffle API to be more flexible for Task. Other shuffle services might also benefit from this way of registration.